elyra/pipeline/component_catalog.py

```diff
@@ -72,90 +70,165 @@ class ComponentCatalog(LoggingConfigurable):
-                             "(ELYRA_COMPONENT_CATALOG_CACHE_TTL env var)").tag(config=True)
-
-    @default('cache_ttl')
-    def cache_ttl_default(self):
-        ttl = ComponentCatalog.ttl_default
-        try:
-            ttl = int(os.getenv(self.cache_ttl_env, ttl))
-        except ValueError:
-            pass
-        return ttl
-
-    def __init__(self,
-                 parser: ComponentParser,
-                 caching_enabled: bool = True,
-                 **kwargs):
+    def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        self._parser = parser
 
-        # Initialize the cache
-        self.caching_enabled = caching_enabled
-        if self.caching_enabled:
-            self.log.debug(f"ComponentCatalog cache TTL: {self.cache_ttl}")
-            self.update_cache()
+        # Indicate whether this is a test environment. This is relevant in
+        # update_cache_for_catalog() - we want to wait for the cache updater
+        # thread to finish in test envs
+        self._for_test = bool(kwargs.get('for_test'))
+
+        self._cache_lock = threading.Lock()
+        if not self._component_cache:
+            self._build_cache()
 
-    def get_all_components(self) -> List[Component]:
+    def _build_cache(self):
         """
-        Retrieve all components; use the component catalog cache if enabled
+        Reads through all catalog Metadata instances to build the initial component cache
         """
-        if self.caching_enabled:
-            if self._is_cache_expired():
-                self.update_cache()
-            return list(self._cached_components.values())
 
-        return list(self._read_component_catalogs().values())
+        all_catalogs = MetadataManager(schemaspace=ComponentCatalogs.COMPONENT_CATALOGS_SCHEMASPACE_ID).get_all()
+        for catalog in all_catalogs:
+            self.update_cache_for_catalog(catalog)
 
-    def get_component(self, component_id: str) -> Optional[Component]:
+    def update_cache_for_catalog(self, catalog: Metadata, operation: Optional[str] = None):
         """
-        Retrieve the component with a given component_id; use component catalog
-        cache if enabled
+        Updates the component cache for the given catalog Metadata instance using a thread.
         """
-        component: Component
-        if self.caching_enabled:
-            if self._is_cache_expired():
-                self.update_cache()
-            component = self._cached_components.get(component_id)
-        else:
-            component = self._read_component_catalogs().get(component_id)
+        platform_type = catalog.metadata['runtime_type']
+
+        # Add sub-dictionary for this platform type if not present
+        if not self._component_cache.get(platform_type):
+            self._component_cache[platform_type] = {}
+
+        def update_cache_with_thread():
+            catalog_components = self._read_component_catalog(catalog)
```

Comment:
This will change when the threading aspect is re-worked, but note that we shouldn't bother to read the catalog if it's a 'delete' operation